src/net/ossl.cc

```diff
@@ -564,7 +564,7 @@ class session : public enable_shared_from_this<session>, public session_impl {
             std::unique_ptr<net::connected_socket_impl> sock, sstring name = { }, std::optional<tls_options> options = std::nullopt)
             : _type(t), _sock(std::move(sock)), _creds(creds->_impl), _hostname(
                    std::move(name)), _in(_sock->source()), _out(_sock->sink()),
-                   _in_sem(1), _out_sem(1),  _options(options.value_or(tls_options{})),
+                   _in_sem(1), _out_sem(1), _send_sem(1), _options(options.value_or(tls_options{})),
```

Comment:
qq: why not just use `_out_sem`?  If the only time you need to write to `_out` is post `BIO_read(_out_bio)` woudln't `_out_sem` suffice?

Reply:
`_out_sem` is taken by `put` and `put` eventually calls `pull_encrypted_and_send`. Therefore there would be a deadlock if I were to guard `_out` with `_out_sem`